### PR TITLE
Make sure certificate is verified when required by RestApiClientAsync

### PR DESCRIPTION
**Pull request by mdazam1942**

### CODE REVIEW

**Review by Rory-Bray — COMMENTED** [review]
Where are the certification verification tests/testcases?

File: stix_shifter_modules/stix_bundle/stix_transmission/connector.py [hunk 1]
```diff
@@ -26,6 +26,7 @@ def __init__(self, connection, configuration):
             auth = BasicAuth(conf_auth['username'], conf_auth['password'])
         self.client = RestApiClientAsync(None,
                                     auth=auth,
+                                    cert_verify=connection.get('selfSignedCert', False),
```
[the remark this answers is not in the document]
Comment (benjamin-craig) [reply stix_shifter_modules/stix_bundle/stix_transmission/connector.py:29]: @mdazam1942  Can you send me the error message that you get when you try to talk to stix bundles using https

[the remark this answers is not in the document]
Comment (delliott90) [reply stix_shifter_modules/stix_bundle/stix_transmission/connector.py:29]: STIX bundles are public URLs, the connector doesn't allow you to even pass a certificate.

[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_modules/stix_bundle/stix_transmission/connector.py:29]: this is a special case. this is basically a demo connector. if anyone wants to use a url that is behind some kind of server that requires ssl cert validation then they can implement in their own. we have test cases using public url that doesn't require any ssl cert validation and in fact it rejects the call if the connector enforces it. here's error:

`stix_shifter_utils.utils.error_response:error_response.py:91 unsupplied connector name connector error occurred: client_connector_error: (Cannot connect to host raw.githubusercontent.com:443 ssl:<ssl.SSLContext object at 0x7fd323d38540> [Network is unreachable])`

**Comment by mdazam1942** [comment]
> Where are the certification verification tests/testcases?

Every connector has functional unittest that run through `RestApiClientAsync` class. The requests/responses are basically mocked in the unittests. For example: https://github.com/opencybersecurityalliance/stix-shifter/blob/c455ab2c7d1fb613d69971c05fce58eddd83b743/stix_shifter_modules/qradar/tests/stix_transmission/test_qradar.py#L21

Can you suggests what kind of test you are looking for?

**Comment by mdazam1942** [comment]
Tested this changes on some connector that uses saas datasources api. They are returning `SSLCertVerificationError`. The reason is, we are enforcing custom SSL certificate validation for all cases even though some datasource doesn't require that. For those api, we need to do default SSL check because the connector doesn't supply certificate. 

We need to revert back few of those changes that require to have certificate

```
if self.ssl_context:
            self.ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS)
            self.ssl_context.verify_mode = ssl.CERT_REQUIRED
            self.ssl_context.check_hostname = True
```

File: stix_shifter_modules/stix_bundle/stix_transmission/connector.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_modules/stix_bundle/stix_transmission/connector.py]: ok so we dont set the cert verify the false anymore. we made sure default ssl check is performed is certificate is not presented.

File: stix_shifter_utils/stix_transmission/utils/RestApiClientAsync.py [hunk 2]
```diff
@@ -87,8 +88,13 @@ async def call_api(self, endpoint, method, headers=None, cookies=None, data=None
...
 
+            if self.ssl_context and self.server_cert_file_content:
                 self.ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS)
-                self.ssl_context.load_verify_locations(self.server_cert_name)
+                try:
+                    self.ssl_context.load_verify_locations(self.server_cert_name)
+                except Exception as ex:
+                    self.logger.debug('Unable to load the certificate for ssl context. Reasons: Connection does not require certificate or unexpected exception while loading the certificate: ' + str(ex))
+
                 self.ssl_context.verify_mode = ssl.CERT_REQUIRED
```
Comment (Rory-Bray) [inline stix_shifter_utils/stix_transmission/utils/RestApiClientAsync.py:97]: verify_mode and check_hostname cannot be conditional they must be applied even in the case of using a default ssl context but I don't see a call to `ssl.create_default_context()` any where.

**Comment by Rory-Bray** [comment]
> > Where are the certification verification tests/testcases?
> 
> Can you suggests what kind of test you are looking for? we can add functional unittests just for this class which will basically mock the ssl verification.

I want to see both positive and negative tests for: trusted and untrusted certificate chains, valid and invalid hostnames, current and expired certs as minimums. Any time you override the default validation you have to explicitly make sure it's actually correctly validating certificates.

**Comment by benjamin-craig** [comment]
> > > Where are the certification verification tests/testcases?
> > 
> > 
> > Can you suggests what kind of test you are looking for? we can add functional unittests just for this class which will basically mock the ssl verification.
> 
> I want to see both positive and negative tests for: trusted and untrusted certificate chains, valid and invalid hostnames, current and expired certs as minimums. Any time you override the default validation you have to explicitly make sure it's actually correctly validating certificates.

@Rory-Bray how do we handle mocking the server side functionality to do these tests?

File: stix_shifter_utils/stix_transmission/utils/RestApiClientAsync.py (hunk 2 above)
Comment (mdazam1942) [reply stix_shifter_utils/stix_transmission/utils/RestApiClientAsync.py]: updated the code to remove ambiguity around certificate validation. now certificate is verified in both custom or default cases. also remove the option to force validation to false.

**Review by Rory-Bray — APPROVED** [review]

**Review by delliott90 — APPROVED** [review]

**Pull request merged**